### PR TITLE
Adds 'handle errors correctly in main' example.

### PR DESCRIPTION
**Pull request by sotocodes**
Issue #214

Note: This example has the category-badge **rust patterns** even though in crates.io error-chain does not belong to any category.

### CODE REVIEW

**Comment by lilianmoraru** [comment]
It is ok, this is the category mentioned in `Cargo.toml`(I guess that's where you got the category from).

**Comment by budziq** [comment]
@somebrel 
> Note: This example has the category-badge rust patterns even though in crates.io error-chain does not belong to any category.

As @lilianmoraru mentioned. You were spot on. `rust patterns` is the category used in latest error-chain master (but not yet pushed to crates.io)

**Comment by lilianmoraru** [comment]
@budziq The category was added 6 months ago: https://github.com/brson/error-chain/blame/master/Cargo.toml#L10
Probably just a `crates.io` bug or something like this.

**Comment by somebrel** [comment]
Perfect 💯

File: src/basics.md [hunk 1]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
 
+[error-handling]: #error-handling
+<a name="error-handling"></a>
+## Handle errors correctly in main
+
+[![error-chain-badge]][error-chain] [![cat-rust-patterns-badge]][cat-rust-patterns]
+
+Uses [error-chain] to handle error occurred when trying to open a file that does not exist.
+
+First, we call `error_chain!{}` inside a new module named  _errors_. This will create the default types `Error`, `ErrorKind`, `ResultExt` and ` Result`, and store them in `errors`. To be able to use these types we must call `use errors::*`.
```
Comment (budziq) [inline src/basics.md]: Please paste identifiers without any suffixes like `()` or `{}` and make sure these are linked to docs like in other examples.

Comment (budziq) [inline src/basics.md]: Please linkify the mentioned identifiers/traits

Comment (budziq) [inline src/basics.md]: > inside a new module named  _errors_. 

Please see my later comment about error module. I would remove it from this basic example and just use `error_chain!` in gloal scope.

File: src/basics.md [hunk 2]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
+
+To handle the erros that may occur when calling `open_foo()`, it must have return type `Reuslt<()>`. We then use [`if let`] when calling it to detect if an error has occurred, and to handle it in an apropiated way.
+```rust
+#[macro_use]
+extern crate error_chain;
+
+mod errors {
+    error_chain!{}
```
Comment (budziq) [inline src/basics.md]: I would define at least a single error within the chain not to use the `chain_err` on str which is convenient but might not be the best pattern to show newbies.

File: src/basics.md [hunk 3]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
+        for e in e.iter().skip(1) {
+            println!("caused by: {}", e);
+        }
+    }
+}
+
+fn open_foo() -> Result<()> {
+    use std::fs::File;
+    File::open("foo.txt")
+        .chain_err(|| "cannot open foo.txt")?;
```
Comment (budziq) [inline src/basics.md]: `chain_err` is an important method. I would mention it in the textual description.

**Comment by budziq** [comment]
> Probably just a crates.io bug or something like this.
Yep you're most likely right :smile:

**Comment by somebrel** [comment]
@budziq  thanks for the feedback, I'll fix it

File: src/basics.md (hunk 2 above)
Comment (somebrel) [inline src/basics.md]: What do you think about this one @budziq?

```rust
#[macro_use]
extern crate error_chain;

error_chain!{
    foreign_links {
        Io(::std::io::Error);
    }   
}

fn main() {
    if let Err(ref e) = read_foo() {
        // handle the error here
        println!("error: {}", e); 
    }   
}

fn read_foo() -> Result<()> {
    use std::fs::File;
    File::open("foo.txt")?;

    Ok(())
}

```

Comment (budziq) [inline src/basics.md]: @somebrel looks good to me :+1: we can defer showing `chain_err` with custom error and error unrolling to a more advanced example. But its your story (I'm just here to help) I am game either way.

I would only drop the `::` prefix

Comment (somebrel) [inline src/basics.md]: alright, thanks @budziq

File: src/basics.md [hunk 4]
```diff
@@ -412,40 +412,38 @@ fn run() -> Result<()> {
...
+<a name="ex-simple-error-handling"></a>
 ## Handle errors correctly in main
 
 [![error-chain-badge]][error-chain] [![cat-rust-patterns-badge]][cat-rust-patterns]
 
-Uses [error-chain] to handle error occurred when trying to open a file that does not exist.
+Handles the error that occur when we try to open a file that does not exist. We do this using [error-chain], a library that takes care of a lot of boilerplate code that an user would have to write in order to handle erros in Rust. For more details please check the chapter [Error Handling] of TRPL.
 
-First, we call `error_chain!{}` inside a new module named  _errors_. This will create the default types `Error`, `ErrorKind`, `ResultExt` and ` Result`, and store them in `errors`. To be able to use these types we must call `use errors::*`.
+The [`error_chain!`] macro creates the types [`Error`], [`ErrorKind`], [`Result`] and [`ResultExt`]. Additionaly, we use `Io(sd::io::Error)` for automatic conversion between this error chain and other error types not defined by [`error_chain!`]. In this case, [`std::io::Error`]
```
Comment (budziq) [inline src/basics.md]: I would avoid going into details of what goes behind the scenes. The reader interested in the internals will go into linked docs. We should keep the noise to minimum with this basic example. Please remove the mention  of " [`Error`], [`ErrorKind`], [`Result`] and [`ResultExt`]. and their links".

Also please see if you can make the description a little shorter

File: src/basics.md [hunk 5]
```diff
@@ -509,8 +506,14 @@ fn open_foo() -> Result<()> {
 [`MutexGuard`]: https://doc.rust-lang.org/std/sync/struct.MutexGuard.html
 [`Mmap::as_slice`]: https://docs.rs/memmap/*/memmap/struct.Mmap.html#method.as_slice
 [`seek`]: https://doc.rust-lang.org/std/fs/struct.File.html#method.seek
-[`if let`]: https://doc.rust-lang.org/book/first-edition/if-let.html
+[`error_chain!`]: https://docs.rs/error-chain/*/error_chain/#declaring-error-types
+[`Error`]: https://docs.rs/error-chain/*/error_chain/example_generated/struct.Error.html
```
Comment (budziq) [inline src/basics.md]: The links with identifiers from std type linking to error_chain generated example are a nice idea in theory but will create confusion in newbie readers as well as pollute the link namespace. I would get rid of them

**Comment by somebrel** [comment]
Thanks for the feeedback @budziq, I can see how the example is getting better.

File: src/basics.md [hunk 6]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
+#[macro_use]
+extern crate error_chain;
+
+error_chain!{
+    foreign_links {
+        Io(std::io::Error);
+    }
+}
+
+fn main() {
```
Comment (budziq) [inline src/basics.md]: I would move the `fn main` to bottom of the example.

File: src/basics.md [hunk 7]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
+
+[![error-chain-badge]][error-chain] [![cat-rust-patterns-badge]][cat-rust-patterns]
+
+Handles error that occur when we try to open a file that does not exist. We do this using [error-chain], a library that takes care of a lot of boilerplate code needed in order to [handle errors in Rust].
+
+We call `Io(std::io::Error)` inside `foreign_links` to convert the type `std::io::Error` into error-chain's custom version of it.
+
+To handle the result of a function that might return an error, in this example `open_foo`, it must return a `Result`. To handle that result, we use pattern matching on it when calling the function. In this example we print an error message if that result was an error.
+
+
```
Comment (budziq) [inline src/basics.md]: I would remove the additional newlines keeping just one

File: src/basics.md [hunk 8]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
+    if let Err(ref e) = read_foo() {
+        // handle the error here
+        println!("error: {}", e);
+    }
+}
+
+fn read_foo() -> Result<()> {
+    use std::fs::File;
+    File::open("foo.txt")?;
+
```
Comment (budziq) [inline src/basics.md]: - I would also add `read_to_string` here to show multiple operations.
- We may even want to actually return the Result<String> and print it in main in `Ok` variant (but I leave it to you)

File: src/basics.md [hunk 9]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
+<a name="ex-simple-error-handling"></a>
+## Handle errors correctly in main
+
+[![error-chain-badge]][error-chain] [![cat-rust-patterns-badge]][cat-rust-patterns]
+
+Handles error that occur when we try to open a file that does not exist. We do this using [error-chain], a library that takes care of a lot of boilerplate code needed in order to [handle errors in Rust].
+
+We call `Io(std::io::Error)` inside `foreign_links` to convert the type `std::io::Error` into error-chain's custom version of it.
+
+To handle the result of a function that might return an error, in this example `open_foo`, it must return a `Result`. To handle that result, we use pattern matching on it when calling the function. In this example we print an error message if that result was an error.
```
Comment (budziq) [inline src/basics.md]: I would avid repetition `To handle the result,` and `To handle that result,`. This might be rephrased and further simplified

File: src/basics.md [hunk 10]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
 
+[ex-simple-error-handling]: #ex-simple-error-handling
+<a name="ex-simple-error-handling"></a>
+## Handle errors correctly in main
+
+[![error-chain-badge]][error-chain] [![cat-rust-patterns-badge]][cat-rust-patterns]
+
+Handles error that occur when we try to open a file that does not exist. We do this using [error-chain], a library that takes care of a lot of boilerplate code needed in order to [handle errors in Rust].
+
+We call `Io(std::io::Error)` inside `foreign_links` to convert the type `std::io::Error` into error-chain's custom version of it.
```
Comment (budziq) [inline src/basics.md]: > We call `Io(std::io::Error)` 

I wouldn't use `call` in this context. It is actually a definition.

> `foreign_links` to convert the type `std::io::Error` into error-chain's custom version of it

I would rephrase it "[`foreign_links`](https://brson.github.io/error-chain/error_chain/index.html#foreign-links) to allow automatic conversion from [`std::io::Error`](https://doc.rust-lang.org/std/io/struct.Error.html) into [`error_chain!`](https://brson.github.io/error-chain/error_chain/macro.error_chain.html) defined type implementing the [`Error`](https://doc.rust-lang.org/std/error/trait.Error.html) trait."

Please note the that I've added 4x links here.

File: src/basics.md (hunk 8 above)
Comment (somebrel) [inline src/basics.md]: Hey @budziq, that's a good Idea... I modified example:
```rust
#[macro_use]
extern crate error_chain;

error_chain!{
    foreign_links {
        Io(std::io::Error);
    }
}

fn read_foo() -> Result<(String)> {
    use std::fs::File;
    use std::io::Read;

    let mut foo = File::open("foo.txt")?;
    let mut foo_content = String::new();
    foo.read_to_string(&mut foo_content)?;
    Ok(foo_content)
}

fn main() {
    match read_foo() {
        Ok(foo_content) => println!("{}", foo_content),
        Err(error) => println!("error: {}", error),
    };
}
```

Comment (budziq) [inline src/basics.md]: Very nice :+1: 

You can skip the `foo` variable, and there is no need to have `String` in parens. Also the use statements would be on top level.
```rust
use std::fs::File;
use std::io::{Result, Read};

fn read_foo() -> Result<String> {
    let mut foo_content = String::new();
    File::open("foo.txt")?.read_to_string(&mut foo_content)?;
    Ok(foo_content)
}

fn main() {
    match read_foo() {
        Ok(foo_content) => println!("{}", foo_content),
        Err(error) => println!("error: {}", error),
    };
}
```

Unfortunately this example can be done fully without any need for error_chain (as presented above)

So we might need to think on a relatively non contrived way to introduce another error variant except  `Io(std::io::Error);` to justify error_chain usage :smile_cat:. You can either:
- define your own app error and return it if foo_content does not match some arbitrary predicate like (starts_with("#"))
- get the filename from env var (and we already have another error variant)
- parse the foo_contents into something like an int
- pass the path by value and manipulate it (for instance strip)
- your idea

File: src/basics.md (hunk 9 above)
Comment (somebrel) [inline src/basics.md]: I changed line 425 to be

 _"open_foo returns a Result, which is the type used for returning and propagating errors. We use pattern matching in main to handle that result, which in our case is either an error or the content of foo.txt."_

What do you thin @budziq?

Comment (budziq) [inline src/basics.md]: sounds ok. please remember to linkify the `Result` identifier.
Also I find the Result definition "the type used for returning and propagating errors" a bit awkward (I know that is taken from official docs but I think that it is lacking and will confuse newbies). What we would like to convey is that:
"Result, which is the type that represents a possibility of failure" (or something along these lines)

**Comment by budziq** [comment]
Hi @Phaiax thanks for all the submissions! 

Unfortunately I'll not be able to follow up with anything resembling review for the next two weeks due to holidays 😞 . I may drop some comments when Im on mobile later and I've asked other maintainers to drop from time to time and look into the PR's.

**Comment by somebrel** [comment]
Hey @budziq, thanks for letting me know. Enjoy your holidays! 😃

**Comment by budziq** [comment]
Hi @somebrel. I'm back from holidays so I should be able to give some more hands on help here. Do you need any assistance with this example?

**Comment by somebrel** [comment]
Welcome back @budziq. I been busy these days but I'll be working on this example Tuesday

**Comment by somebrel** [comment]
Hey @budziq  I went for the application error because I think it is a common one. 

```rust
#[macro_use]
extern crate error_chain;

use std::io::BufReader;
use std::io::BufRead;
use std::fs::File;

error_chain!{
    errors {
        ConfigHeaderMissing {
           description("Invalid header.")
           display("Invalid header. Please check foo.config, line 1")
        }
    }
}

fn check_header() -> Result<bool> {
    let f = File::open("foo.config").unwrap();
    let file = BufReader::new(&f);

    if "#foo.config" == file.lines().next().unwrap().unwrap() {
        Ok(true)
    } else {
        Err(ErrorKind::ConfigHeaderMissing.into())
    }
}

fn main() {
    match check_header() {
        Ok(status) => println!("Valid header: {}", status),
        Err(error) => println!("Error: {}", error),
    };
}

**Comment by budziq** [comment]
Hi @somebrel,

Nice idea but I would use it on one of the more advanced examples for error_chain (there are two more progressing in complexity) also we will need to cleanup `check_header` a little (turn `unwrap`s into `?`s , more natural parsing with `nth` or just compare some [env var](https://doc.rust-lang.org/std/env/fn.var.html) with a set of valid values instead).

But for now I would cleanup the example that we have discussed here previously as it was almost perfect.

**Comment by somebrel** [comment]
alright @budziq  I am back to the previous example.

When I try to parse the content of `foo` into an  `i32` I get `ParseIntError { kind: InvalidDigit }`. The problem is that the content of `foo` is not `12345` but `12345\n`... I am not sure how I use error_chain to customize this error msg :/

**Comment by budziq** [comment]
> I am back to the previous example.

Lets keep the new one ready for next example.

> The problem is that the content of foo is not 12345 but 12345\n... 

You'll have to use `trim` to remove whitespace. 

> I am not sure how I use error_chain to customize this error msg :/

No need to customize the error message in this first example. Let's just handle the standard error correctly.

**Comment by budziq** [comment]
Hi @sotocodes, do you need any help with finishing the work here?

File: src/basics.md [hunk 11]
```diff
@@ -411,6 +412,45 @@ fn run() -> Result<()> {
...
+error_chain!{
+    foreign_links {
+        Io(std::io::Error);
+    }
+}
+
+fn main() {
+    if let Err(ref e) = read_foo() {
+        // handle the error here
+        println!("error: {}", e);
```
Comment (lilianmoraru) [inline src/basics.md]: This should use `eprintln!()`

Comment (lilianmoraru) [inline src/basics.md]: Also, I prefer the `quick_main!` macro(it also handles writing to `stderr` and panicking when it can't), but I'll let @budziq decide here if the explicit error handling is more appropriate here

Comment (budziq) [inline src/basics.md]: > This should use eprintln!()

Good point

> Also, I prefer the quick_main! macro(it also handles writing to stderr and panicking when it can't)

I would opt for explicit error handling this time for educational purposes as this aims to be a super basic example. Reader may refer to `quick_main!` usage in nearly every other example.

I would envision for the main here to mirror (but using `eprintln!`)
https://github.com/dtolnay/cargo-expand/blob/b05fcdabf48fa34c90343b1baa8777b9e65684db/src/main.rs#L18

which pretty much boils down to what `quick_main!` expands to.

**Comment by budziq** [comment]
Hi @sotocodes do you need help in going further with this example?

**Comment by sotocodes** [comment]
Hi @budziq, I have just a few days to finish my thesis. After that I'll try to finish it... sorry for the delay :/

**Comment by budziq** [comment]
>  sorry for the delay :/

No problem man. Just checking if you needed any help. Good luck!

**Comment by budziq** [comment]
Hi @sotocodes are you ready to push this though the finish line?

**Comment by sotocodes** [comment]
Hi @budziq, I haven't been able to code Rust in a while and I'm not sure when I'll be back. So I won't be able to finish these one. Sorry :(

**Comment by budziq** [comment]
@sotocodes No worries mate! I wouldn't want for all of this good work go to waste! 
So if you'll give your consent I would like to add the final touches myself and merge your PR in the next few days (the contribution will stil be yours). Are you ok with that?

Also this would make you a full fledged  contributor requiring your vote on upcoming license change to CC0. So please consider giving also [your consent to re-licensing here](https://github.com/rust-lang-nursery/rust-cookbook/issues/209)

**Comment by sotocodes** [comment]
Sure, it would be great if you could finish it :) 

Thanks

**Comment by ludwigpacifici** [comment]
Hey @budziq, If I want to finish this PR, how should I proceed?

**Comment by budziq** [comment]
It would be super cool If you'd get this through the finish line @ludwigpacifici :+1: 

> If I want to finish this PR, how should I proceed?

1. I'd git pull this PR locally with `git fetch upstream pull/234/head`
2. rebase it to latest master (possibly solving some minor conflicts)
3. then commit and push the work to a separate branch on your fork and make a PR of your own (unfortunately no one except for the author is able to update this PR) marking it as work in progress and linking to the old PR
4. address the remaining [review comments to this PR](https://github.com/rust-lang-nursery/rust-cookbook/pull/234/files) and push the update to your PR with separate commit. 

If you'll have any questions or problems please do not hesitate to ask me :)

**Comment by ludwigpacifici** [comment]
Hey @budziq !

Apologize for the wait and thanks for the detailed reply. I have some free time. I'll have a look!
